File: algorithms/cpp/removeAllAdjacentDuplicatesInString/removeAllAdjacentDuplicatesInString.cpp
// Source: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
// Author: Miao Zhang
// Date:   2021-04-06

class Solution {
public:
    string removeDuplicates(string S) {
        stack<char> st;
        for (char c: S) {
            if (!st.empty() && st.top() == c) st.pop();
            else st.push(c);
        }
        string res = "";
        while(!st.empty()) {
            res = st.top() + res;
            st.pop();
        }
        return res;
    }
};

File: algorithms/python/removeAllAdjacentDuplicatesInString/removeAllAdjacentDuplicatesInString.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
# Author: Miao Zhang
# Date:   2021-04-06

class Solution:
    def removeDuplicates(self, S: str) -> str:
        stack = []
        for c in S:
            if stack and stack[-1] == c:
                stack.pop()
            else:
                stack.append(c)
        return ''.join(stack)
